src/main/java/org/eclipse/pathfinder/api/GraphTraversalService.java

```diff
@@ -6,6 +6,13 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
```

Comment:
I know that Jackson is the more popular API for dealing with JSON in Java. Even so, this is a Jakarta EE sample app and we have a built-in API for JSON in Jakarta EE.  Here's some AI prompt and response.

### AI Prompt

Consider these Java import statements:

```java
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
```

These are from the Jackson JSON processing API. These imports are used
in this Java code:

```java
    @GET
    @Path("/shortest-path")
    @Produces({"application/json", "application/xml; qs=.75"})
    public List<TransitPath> findShortestPath(
                                              @NotNull(message = "Missing origin UN location code.")
                                              @Pattern(
                                                       regexp = "[a-zA-Z]{2}[a-zA-Z2-9]{3}",
                                                       message = "Origin " + UNLOCODE_PATTERN_VIOLATION_MESSAGE)
                                              @QueryParam("origin")
                                              String originUnLocode,
                                              @NotNull(message = "Missing destination UN location code.")
                                              @Pattern(
                                                       regexp = "[a-zA-Z]{2}[a-zA-Z2-9]{3}",
                                                       message = "Destination " + UNLOCODE_PATTERN_VIOLATION_MESSAGE)
                                              @QueryParam("destination")
                                              String destinationUnLocode,
                                              // TODO [DDD] Apply regular expression validation.
                                              @Size(min = 8, max = 8, message = "Deadline value must be eight characters long.")
                                              @QueryParam("deadline")
                                              String deadline) throws JsonProcessingException {
        if (!System.getenv("AZURE_OPENAI_ENDPOINT").isEmpty()
                && !System.getenv("AZURE_OPENAI_KEY").isEmpty()) {
            String shortestPath = getShortestPathWithTimeout(originUnLocode, destinationUnLocode);
            if (isValidJsonUsingJackson(shortestPath) && !shortestPath.equals("[]")) {
                ObjectMapper objectMapper = new ObjectMapper();
                objectMapper.registerModule(new JavaTimeModule());
                List<TransitPath> transitPaths = objectMapper.readValue(shortestPath, new TypeReference<>() {
                });
                if (transitPaths != null) {
                    return transitPaths;
                }
            }
        }

        List<String> allVertices = dao.listLocations();
        allVertices.remove(originUnLocode);
        allVertices.remove(destinationUnLocode);

        int candidateCount = getRandomNumberOfCandidates();
        List<TransitPath> candidates = new ArrayList<>(candidateCount);

        for (int i = 0; i < candidateCount; i++) {
            allVertices = getRandomChunkOfLocations(allVertices);
            List<TransitEdge> transitEdges = new ArrayList<>(allVertices.size() - 1);
            String fromUnLocode = originUnLocode;
            LocalDateTime date = LocalDateTime.now();

            for (int j = 0; j <= allVertices.size(); ++j) {
                LocalDateTime fromDate = nextDate(date);
                LocalDateTime toDate = nextDate(fromDate);
                String toUnLocode = (j >= allVertices.size() ? destinationUnLocode : allVertices.get(j));
                transitEdges.add(
                                 new TransitEdge(
                                                 dao.getVoyageNumber(fromUnLocode, toUnLocode),
                                                 fromUnLocode,
                                                 toUnLocode,
                                                 fromDate,
                                                 toDate));
                fromUnLocode = toUnLocode;
                date = nextDate(toDate);
            }
            candidates.add(new TransitPath(transitEdges));
        }
        return candidates;
    }
```

I am aware there is another API for JSON processing that is defined within Jakarta EE: JSON-B and JSON-P.

Rewrite the import statments and the code to use the Jakarta Standard JSON-P and JSON-B API.

### AI Response

Sure! Here is the updated code using Jakarta JSON-B and JSON-P APIs:

### Import Statements
Replace the Jackson imports with the following:

```java
import jakarta.json.bind.Jsonb;
import jakarta.json.bind.JsonbBuilder;
import jakarta.json.bind.JsonbException;
import jakarta.json.bind.annotation.JsonbDateFormat;
import jakarta.json.bind.annotation.JsonbProperty;
import jakarta.json.bind.spi.JsonbProvider;
import jakarta.json.Json;
import jakarta.json.JsonArray;
import jakarta.json.JsonObject;
import jakarta.json.JsonReader;
import jakarta.json.JsonValue;
import jakarta.json.spi.JsonProvider;
```

### Updated Code
Here is the updated method using Jakarta JSON-B and JSON-P:

```java
@GET
@Path("/shortest-path")
@Produces({"application/json", "application/xml; qs=.75"})
public List<TransitPath> findShortestPath(
                                          @NotNull(message = "Missing origin UN location code.")
                                          @Pattern(
                                                   regexp = "[a-zA-Z]{2}[a-zA-Z2-9]{3}",
                                                   message = "Origin " + UNLOCODE_PATTERN_VIOLATION_MESSAGE)
                                          @QueryParam("origin")
                                          String originUnLocode,
                                          @NotNull(message = "Missing destination UN location code.")
                                          @Pattern(
                                                   regexp = "[a-zA-Z]{2}[a-zA-Z2-9]{3}",
                                                   message = "Destination " + UNLOCODE_PATTERN_VIOLATION_MESSAGE)
                                          @QueryParam("destination")
                                          String destinationUnLocode,
                                          @Size(min = 8, max = 8, message = "Deadline value must be eight characters long.")
                                          @QueryParam("deadline")
                                          String deadline) throws JsonbException {
    if (!System.getenv("AZURE_OPENAI_ENDPOINT").isEmpty()
            && !System.getenv("AZURE_OPENAI_KEY").isEmpty()) {
        String shortestPath = getShortestPathWithTimeout(originUnLocode, destinationUnLocode);
        if (isValidJsonUsingJsonP(shortestPath) && !shortestPath.equals("[]")) {
            Jsonb jsonb = JsonbBuilder.create();
            List<TransitPath> transitPaths = jsonb.fromJson(shortestPath, new ArrayList<TransitPath>(){}.getClass().getGenericSuperclass());
            if (transitPaths != null) {
                return transitPaths;
            }
        }
    }

    List<String> allVertices = dao.listLocations();
    allVertices.remove(originUnLocode);
    allVertices.remove(destinationUnLocode);

    int candidateCount = getRandomNumberOfCandidates();
    List<TransitPath> candidates = new ArrayList<>(candidateCount);

    for (int i = 0; i < candidateCount; i++) {
        allVertices = getRandomChunkOfLocations(allVertices);
        List<TransitEdge> transitEdges = new ArrayList<>(allVertices.size() - 1);
        String fromUnLocode = originUnLocode;
        LocalDateTime date = LocalDateTime.now();

        for (int j = 0; j <= allVertices.size(); ++j) {
            LocalDateTime fromDate = nextDate(date);
            LocalDateTime toDate = nextDate(fromDate);
            String toUnLocode = (j >= allVertices.size() ? destinationUnLocode : allVertices.get(j));
            transitEdges.add(
                             new TransitEdge(
                                             dao.getVoyageNumber(fromUnLocode, toUnLocode),
                                             fromUnLocode,
                                             toUnLocode,
                                             fromDate,
                                             toDate));
            fromUnLocode = toUnLocode;
            date = nextDate(toDate);
        }
        candidates.add(new TransitPath(transitEdges));
    }
    return candidates;
}

private boolean isValidJsonUsingJsonP(String json) {
    try (JsonReader reader = Json.createReader(new StringReader(json))) {
        JsonValue jsonValue = reader.read();
        return jsonValue.getValueType() == JsonValue.ValueType.ARRAY;
    } catch (Exception e) {
        return false;
    }
}
```

This code uses Jakarta JSON-B for binding JSON data to Java objects and JSON-P for processing JSON data. The `isValidJsonUsingJsonP` method checks if the JSON string is a valid JSON array.

Source: Conversation with Copilot, 10/3/2024
(1) github.com. https://github.com/Azure-Samples/cargotracker-azure/tree/d2305ee5ba1990c1594789b740a08e73fcab8580/src%2Fmain%2Fjava%2Forg%2Feclipse%2Fpathfinder%2Fapi%2FGraphTraversalService.java.
(2) github.com. https://github.com/Dimitrios79/Cargo-tracker/tree/c9dc0724954d085b210311eb3287cafab809cfd5/src%2Fmain%2Fjava%2Fnet%2Fjava%2Fpathfinder%2Fapi%2FGraphTraversalService.java.